nntrainer/dataset/func_data_producer.cpp

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file   func_data_producer.cpp
+ * @date   12 July 2021
+ * @brief  This file contains various data producers from a callback
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+
+#include <func_data_producer.h>
+
+#include <nntrainer_error.h>
+
+namespace nntrainer {
+
+FuncDataProducer::FuncDataProducer(datagen_cb datagen_cb, void *user_data_) :
+  cb(datagen_cb),
+  user_data(user_data_) {}
+
+FuncDataProducer::~FuncDataProducer() {}
+
+const std::string FuncDataProducer::getType() const {
+  return FuncDataProducer::type;
+}
+
+void FuncDataProducer::setProperty(const std::vector<std::string> &properties) {
+  NNTR_THROW_IF(!properties.empty(), std::invalid_argument)
+    << "properties is not empty, size: " << properties.size();
+}
+
+DataProducer::Gernerator
+FuncDataProducer::finalize(const std::vector<TensorDim> &input_dims,
+                           const std::vector<TensorDim> &label_dims) {
+  return [cb = this->cb, ud = this->user_data, input_dims,
+          label_dims]() -> DataProducer::Iteration {
+    std::vector<Tensor> inputs;
+    inputs.reserve(input_dims.size());
+    float **input_data = new float *[input_dims.size()];
```

Comment:
can you create this as a member of this class than being created at every call?

Reply:
also the corresponding free is missing.

Reply:
Yes, I changed to unique_ptr and member of this lambda